
# Czyszczenie danych Titanic


## Celem naszego zadania jest analiza zbioru danych "TitanicMess.tsv", zidentyfikowanie problemów, oczyszczenie danych.

Ostatecznym punktem zadania jest ukazanie danych po analizie 

Pracę rozpoczynam od zaimportowania niezbędnych bibliotek oraz importu danych

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math as math
import statistics


%matplotlib inline

Zaimportowanie danych 

In [4]:
titanic_dt = pd.read_csv('C:/Users/user/Desktop/TitanicMess.tsv', sep = '\t')
titanic_dt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1   
2                             Heikkinen, Miss. Laina  female  26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1   
4                           Allen, Mr. William Henry    male  35      0   

   Parch            Ticket     Fare Cabin Embarked     ship  
0      0         A/5 21171     7,25   NaN        S  Titanic  
1      0          PC 17599  71,2833   C85        C  Titanic  
2      0  STON/O2. 3101282    7,925   NaN        S  Titanic  
3      0            113803     53,1  C123        S  Titanic  
4      0            373450     8,05   NaN        S  Titanic

Korzystam z funkcji head do wyświetlenia 5 pierwszych wierszy z ramki danych titanic_dt

Używam funkcji "tail" do wyświetlenia końcowej sekwencji danych

In [5]:
titanic_dt.tail()

PassengerId  Survived  Pclass                                      Name  \
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   
891         1000         1       1               Mr. Frederick Maxfield Hoyt   

        Sex  Age  SibSp  Parch      Ticket   Fare Cabin Embarked     ship  
887  female   19      0      0      112053     30   B42        S  Titanic  
888  female  NaN      1      2  W./C. 6607  23,45   NaN        S  Titanic  
889    male   26      0      0      111369     30  C148        C  Titanic  
890    male   32      0      0      370376   7,75   NaN        Q  Titanic  
891    male   38      1      0       19943     90   C93        S  Titanic

# Zauważamy kilka kwestii:
a) W zbiorze występują kolumny w których znajdują się brakujące dane. Na dwóch powyższych wycinkach są one w kolumnie "Age" i "Cabin".

b) Wiele kolumn posiada dane kategoryzujące. Nominalne (kolumna Sex) oraz porządkowe (kolumna Pclass). Dane przedstawione w postaci numerycznej oraz jako łańcuch znaków.

c) Dane numeryczne nie są przeskalowane - dotyczy to np kolumny Age.

d) Dane mogą się duplikować.

e) Do analizy chcę wybrać kolumnę, której wartość chcę przewidywać, oraz wyselekcjonować odpowiednie kolumny niezbędne do danego przewidywania.


Jak zauważamy kolumna Cabin nie dostarcza nam żadnej istotnej informacji do analizy, wobec tego usuwam ją z ramki danych

In [6]:
titanic_dt.drop('Cabin',axis=1, inplace=True)

taki sam zabieg wykonuję dla kolumny "Embarked"

In [7]:
titanic_dt.drop('Embarked',axis=1, inplace=True)

Brakujące dane- wyświetlenie ich wartości 

In [8]:
column_names = titanic_dt.columns.tolist()
for column in column_names:
    print(column)
    print(titanic_dt[column].value_counts(dropna = False))

PassengerId
225    3
11     3
295    1
306    1
305    1
      ..
592    1
591    1
590    1
589    1
1      1
Name: PassengerId, Length: 888, dtype: int64
Survived
0    547
1    345
Name: Survived, dtype: int64
Pclass
3    491
1    217
2    184
Name: Pclass, dtype: int64
Name
Sandstrom, Miss. Marguerite Ru&5$$    3
Hoyt, Mr. Frederick Maxfield          2
Windelov, Mr. Einar                   1
Dowdell, Miss. Elizabeth              1
Scanlan, Mr. James                    1
                                     ..
Olsen, Mr. Karl Siegwart Andreas      1
Pain, Dr. Alfred                      1
Sirayanian, Mr. Orsen                 1
Frauenthal, Dr. Henry William         1
Wiklund, Mr. Jakob Alfred             1
Name: Name, Length: 889, dtype: int64
Sex
male       576
female     312
fem          1
malef        1
femmale      1
mal          1
Name: Sex, dtype: int64
Age
NaN     173
24       30
22       27
18       26
19       25
       ... 
70,5      1
14,5      1
12        1
250       1
55

Obserwujemy, że brakuje wartości w 173 wierszach w kolumnie Age oraz wartości których nie są skalowalne.

Cztery wartości w kolumnie Sex nie mają odpowiedniej nazwy przez co nie są dobrze przypisane do wartości.

In [9]:
titanic_dt['Age'].isnull().values.any()
titanic_dt['Age'].isnull().sum()

173

Tworzę tabelę z wybranymi agrumentami 

In [10]:
titanic_dt = titanic_dt[['PassengerId','Survived','Sex','Age','SibSp','Parch','Fare']]

Następnie usuwam wszystkie wartości'N/A' z tabeli

In [11]:
def drop_na(titanic_n):
    
    titanic_dt = len(titanic_n)
    titanic_n = titanic_n.dropna()
    print("Brakujące wartości usunięte: {}".format(titanic_dt - len(titanic_n)))
    
titanic_n = drop_na(titanic_dt)    

Brakujące wartości usunięte: 173


To samo robię z duplikatami

In [12]:
def drop_dupl(titanic_n):
    
    duplicate = len(titanic_n[titanic_n['PassengerId'].duplicated()])
    if duplicate > 0:
        titanic_n = titanic_n.drop_duplicates(subset = 'PassengerId', keep="last")
        print ("Usunięte duplikaty: {}".format(duplicate))
        return titanic_n
    else:
        return titanic_n
    
titanic_n = drop_dupl(titanic_dt)

Usunięte duplikaty: 4


Sprawdzam czy nadal występują duplikaty

In [13]:
titanic_n[titanic_n.duplicated()]

Empty DataFrame
Columns: [PassengerId, Survived, Sex, Age, SibSp, Parch, Fare]
Index: []

In [14]:
titanic_n.dropna(inplace= True)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Robię porządek w kolumnie Sex ze względu na błędne wartości dotyczące nazwy płci. 

In [15]:
titanic_n = titanic_n.copy()
titanic_n['Sex'] = titanic_n['Sex']. replace({'malef': 'male','mal': 'male','femmale': 'female','fem': 'female'})

In [16]:
print(titanic_n['Sex'].unique())

['male' 'female']


## Weryfikuję wyczyszczone dane 

In [17]:
titanic_n.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 715 entries, 0 to 891
Data columns (total 7 columns):
PassengerId    715 non-null int64
Survived       715 non-null int64
Sex            715 non-null object
Age            715 non-null object
SibSp          715 non-null int64
Parch          715 non-null int64
Fare           715 non-null object
dtypes: int64(4), object(3)
memory usage: 44.7+ KB


Sprawdzam dokładniej kolumnę Age ze względu na wcześniej wyłapane w niej błędy

In [18]:
titanic_n.Age.unique()

array(['22', '38', '26', '35', '54', '2', '27', '14', '58', '20', '55',
       '31', '34', '15', '4', '8', '19', '40', '.9', '66', '28', '42',
       '21', '18', '3', '7', '49', '29', '65', '28,5', '5', '11', '45',
       '17', '32', '16', '25', '0,83', '30', '33', '23', '24', '46', '59',
       '71', '37', '47', '14,5', '70,5', '32,5', '12', '9', '36,5', '51',
       '55,5', '40,5', '44', '1', '61', '56', '50', '36', '45,5', '4435',
       '20,5', '62', '41', '52', '63', '23,5', '0,92', '43', '60', '39',
       '10', '64', '13', '48', '0,75', '-3', '-12', '53', '57', '80',
       '250', '70', '24,5', '6', '0,67', '30,5', '0,42', '34,5', '74'],
      dtype=object)

In [19]:
newAge= titanic_n['Age'].value_counts().to_dict()

In [20]:
display(newAge)

{'24': 30,
 '22': 27,
 '18': 26,
 '19': 25,
 '21': 24,
 '30': 24,
 '28': 24,
 '36': 22,
 '25': 22,
 '29': 20,
 '35': 18,
 '32': 18,
 '26': 18,
 '27': 18,
 '16': 17,
 '31': 17,
 '23': 15,
 '33': 15,
 '34': 15,
 '20': 15,
 '40': 13,
 '17': 13,
 '39': 13,
 '42': 13,
 '45': 12,
 '38': 12,
 '2': 10,
 '4': 10,
 '50': 10,
 '47': 9,
 '48': 9,
 '44': 9,
 '54': 8,
 '9': 8,
 '51': 7,
 '1': 7,
 '49': 6,
 '3': 6,
 '52': 6,
 '37': 6,
 '41': 6,
 '14': 6,
 '15': 5,
 '58': 5,
 '60': 4,
 '8': 4,
 '5': 4,
 '62': 4,
 '56': 4,
 '11': 4,
 '43': 4,
 '46': 3,
 '6': 3,
 '65': 3,
 '61': 3,
 '7': 3,
 '71': 2,
 '70': 2,
 '63': 2,
 '45,5': 2,
 '28,5': 2,
 '0,75': 2,
 '32,5': 2,
 '30,5': 2,
 '59': 2,
 '64': 2,
 '57': 2,
 '10': 2,
 '0,83': 2,
 '13': 2,
 '40,5': 2,
 '55': 2,
 '70,5': 1,
 '0,92': 1,
 '36,5': 1,
 '20,5': 1,
 '14,5': 1,
 '74': 1,
 '-3': 1,
 '34,5': 1,
 '-12': 1,
 '250': 1,
 '.9': 1,
 '12': 1,
 '53': 1,
 '66': 1,
 '0,67': 1,
 '24,5': 1,
 '4435': 1,
 '80': 1,
 '0,42': 1,
 '23,5': 1,
 '55,5': 1}

Zauważamy że w kolumnie Age występują wartości dziesiętne, ujemne, które decyduję się usunąć.

In [21]:
titanic_n = titanic_n[titanic_n.Age.apply(lambda x: x.isnumeric())]

In [22]:
titanic_n

PassengerId  Survived     Sex Age  SibSp  Parch     Fare
0              1         0    male  22      1      0     7,25
1              2         1  female  38      1      0  71,2833
2              3         1  female  26      0      0    7,925
3              4         1  female  35      1      0     53,1
4              5         0    male  35      0      0     8,05
..           ...       ...     ...  ..    ...    ...      ...
886          887         0    male  27      0      0       13
887          888         1  female  19      0      0       30
889          890         1    male  26      0      0       30
890          891         0    male  32      0      0     7,75
891         1000         1    male  38      1      0       90

[687 rows x 7 columns]

In [23]:
titanic_n= titanic_n.astype({"Age": int, "Survived": bool})

In [24]:
print(titanic_n[titanic_n['Age'] > 100])
titanic_n = titanic_n[titanic_n ['Age']< 100]

     PassengerId  Survived   Sex   Age  SibSp  Parch    Fare
223          224     False  male  4435      0      0  7,8958
666          667     False  male   250      0      0      13


In [ ]:
#zakładam, iż wiek powyżej 100 lat nie jest brany pod uwagę w naszej analizie, dlatego wartości nieprawidłowe powyżej 100 zostają usunięte

Plik został oczyszczony zgodznie z założonymi przeze mnie punktami.
Zapisuję go jako plik :TitanicCleaned.tsv"

In [25]:
titanic_n.to_csv('TitanicCleaned.tsv', index = False)

## Wnioski ##

Wgrany przez nas plik do analizy "TitanicMess.tsv" wymagał usunięcia duplikatów, wartości brakujących oraz zbędnych kolumn. Niektóre wartości wymagały również delikatnej poprawy pisowni. 
Brakujące wartości zdecydowałam się usunąć ponieważ było ich zbyt wiele i w większości dotyczyły one atrybutów które nie miałyby większego znaczenia również ograniczenie  dalszej analizie danych. Bardzo korzystnym posunięciem było również ograniczenie zbioru do kolumn PassengerId, Survived, Sex, Age, SibSp, Parch, Fare. 

Powyższy skrypt może być wykorzystany dla innych danych wejściowych, ale z wcześniejszym zapoznaniem się z danymi, ponieważ w powyższej analizie zostały wykonane czynności, które przy analizie innych danych nie powinny/ nie musiałyby mieć miejsca jak np. poprawa literówek dla kolumny Sex.  